https://www.kaggle.com/datasets/mehmetsabrikunt/internet-service-churn

There is a big competition between Internet providers. 
If a providers want to increase its revenue they needs more subscriber but keep existing customer is more important than having new ones. 
So providers want to know which customer should cancel his service. 
we call this as churn. if the know who will go, maybe they can catch them with promotions.

project Goals:

project Requirements:



Industry: Telecommunications

Aim: This task is aimed at predicting customers behavior in the telecommuniation.

Task: Predict if a customer will unsubscribe and create churn or not (classification and regression)

Dataset: Structured telecom internet subscriber data. the data set is made up of 72274 rows, and 11 columns

Type of Learning: Supervised learning
Models: Logistic regression, random forest classification
Output: Yes or No if customer churn is predicted

Use Case
Our use case for this project was the Customer Churn Dataset sourced from kaggle
(https://www.kaggle.com/datasets/mehmetsabrikunt/internet-service-churn) directory. 

Objectives
Our objective is to determine whether a customer will leave the service provider and what could be the factor that sorround the customers behavior.

We will fit 2 classification models and find the best model to describe our data. The models are:
        Random Forest
        Logistic Regression
        
Data Cleaning



                        ***** DATA DESCRIPTION *****

Our data set has 72274 rows and 11 columns. Each entry contains the following information;

       1. Id:This is customers unique identification number.

       2. is_tv_subcriber: These are customers that subscribes to tv packages
        
       3. is_movie_package_subscriber: Customers that subscribes to movie packages

       4. subscription_age: Age of the subscribers.
        
       5. bill_avg: customers billing average.
        
       6. reamining_contract: 
        
       7. service_failure_count
       8. download_avg

       9. upload_avg
        
      10. download_over_limit
        
      11. churn: describes the loss of customers who don't resign their contract at the time of their renewal.

In [1]:
#importing the necesassary library
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [4]:
df = pd.read_csv("data/internet_service_churn.csv")

In [3]:
telecom.head(2)


id  is_tv_subscriber  is_movie_package_subscriber  subscription_age  \
0  15                 1                            0             11.95   
1  18                 0                            0              8.22   

   bill_avg  reamining_contract  service_failure_count  download_avg  \
0        25                0.14                      0           8.4   
1         0                 NaN                      0           0.0   

   upload_avg  download_over_limit  churn  
0         2.3                    0      0  
1         0.0                    0      1

In [4]:
telecom.tail()

id  is_tv_subscriber  is_movie_package_subscriber  \
72269  1689648                 1                            1   
72270  1689676                 1                            0   
72271  1689679                 1                            0   
72272  1689733                 0                            0   
72273  1689744                 1                            1   

       subscription_age  bill_avg  reamining_contract  service_failure_count  \
72269              0.09         0                1.25                      0   
72270              0.06         1                1.63                      0   
72271              0.02         0                2.19                      0   
72272              0.01         0                0.72                      0   
72273              0.01         0                0.82                      0   

       download_avg  upload_avg  download_over_limit  churn  
72269           0.0         0.0                    0      1  
72270           0.8         0.0                    0      1  
72271           1.5         0.2                    0      1  
72272           0.0         0.0                    0      1  
72273           0.0         0.0                    0      1

In [5]:
telecom.shape

(72274, 11)

In [6]:
#checking for null values in the dataset.
telecom.isnull().any()

id                             False
is_tv_subscriber               False
is_movie_package_subscriber    False
subscription_age               False
bill_avg                       False
reamining_contract              True
service_failure_count          False
download_avg                    True
upload_avg                      True
download_over_limit            False
churn                          False
dtype: bool

    Observation:
There are three columns that contains null values which are; reamining_contract, download_avg and upload_avg 

In [7]:
#checking the data type of the data set (columns)
telecom.dtypes

id                               int64
is_tv_subscriber                 int64
is_movie_package_subscriber      int64
subscription_age               float64
bill_avg                         int64
reamining_contract             float64
service_failure_count            int64
download_avg                   float64
upload_avg                     float64
download_over_limit              int64
churn                            int64
dtype: object

In [8]:
# identifying and removing null values
telecom['download_avg'].replace('', np.nan, inplace=True)
telecom['upload_avg'].replace('', np.nan, inplace=True)
telecom['reamining_contract'].replace('', np.nan, inplace=True)

telecom.dropna(subset=['reamining_contract','download_avg', 'upload_avg'], inplace=True)


In [9]:
telecom.head()

id  is_tv_subscriber  is_movie_package_subscriber  subscription_age  \
0    15                 1                            0             11.95   
2    23                 1                            0              8.91   
5    56                 1                            1             11.94   
6    71                 0                            0              8.96   
10  166                 1                            1              8.44   

    bill_avg  reamining_contract  service_failure_count  download_avg  \
0         25                0.14                      0           8.4   
2         16                0.00                      0          13.7   
5         32                1.38                      0          69.4   
6         18                0.00                      0          21.3   
10        17                0.00                      0           9.9   

    upload_avg  download_over_limit  churn  
0          2.3                    0      0  
2          0.9                    0      1  
5          4.0                    0      0  
6          2.0                    0      1  
10         1.1                    0      1

In [10]:
telecom['is_tv_subscriber'].value_counts()

1    45029
0     5346
Name: is_tv_subscriber, dtype: int64

In [11]:
telecom['is_movie_package_subscriber'].unique()


array([0, 1])

In [12]:
telecom['is_movie_package_subscriber'].value_counts()

0    30307
1    20068
Name: is_movie_package_subscriber, dtype: int64

zero (0) and one (1) are used to represent the non subscribers and the subscribers.

We have two subcribers, namely; 
the tv_subscriber and the movie_package_subscriber. 
After having droped the null values, We have 45029 total number of tv subcribers and 5346 non subcribers while movie package subscriber are 20068 and non subscriber are 30307.

From here, the higher customers are in tv subscription than movie subcription.

In [13]:
telecom.head(3)

id  is_tv_subscriber  is_movie_package_subscriber  subscription_age  \
0  15                 1                            0             11.95   
2  23                 1                            0              8.91   
5  56                 1                            1             11.94   

   bill_avg  reamining_contract  service_failure_count  download_avg  \
0        25                0.14                      0           8.4   
2        16                0.00                      0          13.7   
5        32                1.38                      0          69.4   

   upload_avg  download_over_limit  churn  
0         2.3                    0      0  
2         0.9                    0      1  
5         4.0                    0      0

In [14]:
telecom.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50375 entries, 0 to 72273
Data columns (total 11 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   id                           50375 non-null  int64  
 1   is_tv_subscriber             50375 non-null  int64  
 2   is_movie_package_subscriber  50375 non-null  int64  
 3   subscription_age             50375 non-null  float64
 4   bill_avg                     50375 non-null  int64  
 5   reamining_contract           50375 non-null  float64
 6   service_failure_count        50375 non-null  int64  
 7   download_avg                 50375 non-null  float64
 8   upload_avg                   50375 non-null  float64
 9   download_over_limit          50375 non-null  int64  
 10  churn                        50375 non-null  int64  
dtypes: float64(4), int64(7)
memory usage: 4.6 MB


In [15]:
telecom.describe()

id  is_tv_subscriber  is_movie_package_subscriber  \
count  5.037500e+04      50375.000000                 50375.000000   
mean   9.823436e+05          0.893876                     0.398372   
std    4.577668e+05          0.307999                     0.489568   
min    1.500000e+01          0.000000                     0.000000   
25%    6.229865e+05          1.000000                     0.000000   
50%    1.029724e+06          1.000000                     0.000000   
75%    1.379053e+06          1.000000                     1.000000   
max    1.689744e+06          1.000000                     1.000000   

       subscription_age      bill_avg  reamining_contract  \
count       50375.00000  50375.000000        50375.000000   
mean            2.30913     18.388189            0.710171   
std             2.10370     10.045695            0.693979   
min            -0.02000      0.000000            0.000000   
25%             0.77000     13.000000            0.000000   
50%             1.61000     18.000000            0.560000   
75%             3.17000     22.000000            1.300000   
max            12.76000    386.000000            2.920000   

       service_failure_count  download_avg    upload_avg  download_over_limit  \
count           50375.000000  50375.000000  50375.000000         50375.000000   
mean                0.275494     51.188812      4.823740             0.075136   
std                 0.807156     59.527979     10.166523             0.564437   
min                 0.000000      0.000000      0.000000             0.000000   
25%                 0.000000     14.600000      1.000000             0.000000   
50%                 0.000000     37.200000      2.800000             0.000000   
75%                 0.000000     70.400000      5.600000             0.000000   
max                18.000000   4415.200000    453.300000             7.000000   

              churn  
count  50375.000000  
mean       0.403593  
std        0.490623  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        1.000000  
max        1.000000

In [16]:
# rows and columns left for our predcition after having droped the null values
telecom.shape

(50375, 11)

In [17]:
#setting index as identifier, thereby making the work more expressive

#Index can be used to extraxt value from the dataframe

telecom.set_index('id', inplace =True)

In [21]:
telecom.shape

(50375, 10)

In [18]:
telecom.head(2)

is_tv_subscriber  is_movie_package_subscriber  subscription_age  bill_avg  \
id                                                                              
15                 1                            0             11.95        25   
23                 1                            0              8.91        16   

    reamining_contract  service_failure_count  download_avg  upload_avg  \
id                                                                        
15                0.14                      0           8.4         2.3   
23                0.00                      0          13.7         0.9   

    download_over_limit  churn  
id                              
15                    0      0  
23                    0      1

In [22]:
# let's get a record of a particular identifier. that is to extract value from the dataset
telecom.loc[23]

is_tv_subscriber                1.00
is_movie_package_subscriber     0.00
subscription_age                8.91
bill_avg                       16.00
reamining_contract              0.00
service_failure_count           0.00
download_avg                   13.70
upload_avg                      0.90
download_over_limit             0.00
churn                           1.00
Name: 23, dtype: float64

In [25]:
#extracting multiple values

telecom.loc[[15,23 ]]

is_tv_subscriber  is_movie_package_subscriber  subscription_age  bill_avg  \
id                                                                              
15                 1                            0             11.95        25   
23                 1                            0              8.91        16   

    reamining_contract  service_failure_count  download_avg  upload_avg  \
id                                                                        
15                0.14                      0           8.4         2.3   
23                0.00                      0          13.7         0.9   

    download_over_limit  churn  
id                              
15                    0      0  
23                    0      1

In [26]:
telecom['download_avg'].unique()


array([  8.4,  13.7,  69.4, ..., 229.6, 192.9, 254.9])

In [27]:
telecom['download_over_limit'].value_counts()

0    49137
1      379
2      239
3      159
4      139
6      127
5      116
7       79
Name: download_over_limit, dtype: int64

In [34]:
telecom['churn'].value_counts()

0    30044
1    20331
Name: churn, dtype: int64

In [42]:
churn1=telecom[telecom['churn']==1]
churn1

is_tv_subscriber  is_movie_package_subscriber  subscription_age  \
id                                                                         
23                      1                            0              8.91   
71                      0                            0              8.96   
166                     1                            1              8.44   
270                     0                            0              6.97   
682                     1                            0             10.93   
...                   ...                          ...               ...   
1689648                 1                            1              0.09   
1689676                 1                            0              0.06   
1689679                 1                            0              0.02   
1689733                 0                            0              0.01   
1689744                 1                            1              0.01   

         bill_avg  reamining_contract  service_failure_count  download_avg  \
id                                                                           
23             16                0.00                      0          13.7   
71             18                0.00                      0          21.3   
166            17                0.00                      0           9.9   
270            23                0.28                      0           0.0   
682            63                0.00                      0          38.1   
...           ...                 ...                    ...           ...   
1689648         0                1.25                      0           0.0   
1689676         1                1.63                      0           0.8   
1689679         0                2.19                      0           1.5   
1689733         0                0.72                      0           0.0   
1689744         0                0.82                      0           0.0   

         upload_avg  download_over_limit  churn  
id                                               
23              0.9                    0      1  
71              2.0                    0      1  
166             1.1                    0      1  
270             0.0                    0      1  
682             7.9                    0      1  
...             ...                  ...    ...  
1689648         0.0                    0      1  
1689676         0.0                    0      1  
1689679         0.2                    0      1  
1689733         0.0                    0      1  
1689744         0.0                    0      1  

[20331 rows x 10 columns]

customers that left are greater that the customers that renewed their subscription.

In [40]:
telecom['subscription_age'].unique()

array([11.95,  8.91, 11.94, ...,  2.26,  2.49, -0.02])

In [5]:
  df['is_contract'] = df['reamining_contract'].apply(lambda ele: 0 if pd.isna(ele) else 1)

In [7]:
df['reamining_contract'].replace(np.nan, 0, inplace=True)
    # Rearranging columns
column_names = ['id', 'is_tv_subscriber', 'is_movie_package_subscriber', 'subscription_age', 'bill_avg',
                'reamining_contract',
                'is_contract', 'service_failure_count', 'download_avg', 'upload_avg', 'download_over_limit',
                'churn']

df = df.reindex(columns=column_names)

In [8]:
df.corr()['churn'].sort_values(ascending=False)

churn                          1.000000
download_over_limit            0.158318
service_failure_count          0.021575
bill_avg                      -0.021058
subscription_age              -0.122441
upload_avg                    -0.162690
download_avg                  -0.298063
is_movie_package_subscriber   -0.307167
is_tv_subscriber              -0.327582
id                            -0.451142
is_contract                   -0.472366
reamining_contract            -0.681611
Name: churn, dtype: float64

In [11]:
telecom = df

In [12]:
telecom.head()

id  is_tv_subscriber  is_movie_package_subscriber  subscription_age  \
0  15                 1                            0             11.95   
1  18                 0                            0              8.22   
2  23                 1                            0              8.91   
3  27                 0                            0              6.87   
4  34                 0                            0              6.39   

   bill_avg  reamining_contract  is_contract  service_failure_count  \
0        25                0.14            1                      0   
1         0                0.00            0                      0   
2        16                0.00            1                      0   
3        21                0.00            0                      1   
4         0                0.00            0                      0   

   download_avg  upload_avg  download_over_limit  churn  
0           8.4         2.3                    0      0  
1           0.0         0.0                    0      1  
2          13.7         0.9                    0      1  
3           0.0         0.0                    0      1  
4           0.0         0.0                    0      1

In [13]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

warnings.filterwarnings('ignore')

In [21]:
 df.corr()['churn'].sort_values(ascending=False)

# Splitting up training and Label features for training
x = df.drop(columns=['churn'])
y = df['churn'].values

# if data_max_size != x.shape[0]:
#     x = x.head(data_max_size)
#     y = y[:data_max_size]
# logger.info('[Data] DataPreparation Time Taken in seconds --> %f secs', time.time() - start_time)
# logger.info('[Data] Total Data samples ---> %s', x.shape[0])

# Preparing dataset for Training
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=40, stratify=y)
# Since the numerical features are distributed over different value ranges,
# standard scalar is used to scale them down to the same range.
num_cols = ['subscription_age', 'reamining_contract', 'download_avg', 'upload_avg',
            'download_over_limit', 'bill_avg', 'service_failure_count']

scaler = StandardScaler()

x_train[num_cols] = scaler.fit_transform(x_train[num_cols])
x_test[num_cols] = scaler.transform(x_test[num_cols])

x_test

id  is_tv_subscriber  is_movie_package_subscriber  \
63409  1486704                 1                            0   
24117   566013                 1                            1   
20603   480982                 1                            0   
44828  1050653                 1                            0   
31428   737336                 1                            0   
...        ...               ...                          ...   
26896   630379                 0                            0   
17365   406109                 1                            1   
53692  1258056                 1                            0   
57525  1347223                 1                            0   
36996   867343                 1                            0   

       subscription_age  bill_avg  reamining_contract  is_contract  \
63409         -0.826823 -0.375039            1.038254            1   
24117          1.637090 -0.299411            1.321616            1   
20603          2.104299  0.229990            0.098684            1   
44828          0.151857  0.229990            0.426788            1   
31428          1.032177  7.717229           -0.751403            0   
...                 ...       ...                 ...          ...   
26896         -0.443220  0.154361           -0.751403            1   
17365          1.922333  0.003104            2.141875            1   
53692         -0.354696 -0.450668            2.067306            1   
57525         -1.082559  0.532505           -0.751403            1   
36996         -0.448138  0.381247           -0.751403            1   

       service_failure_count  download_avg  upload_avg  download_over_limit  
63409              -0.332086      0.552304    0.020135            -0.208516  
24117              -0.332086     -0.270605    0.811830            -0.208516  
20603              -0.332086      0.388657    0.060735            -0.208516  
44828               0.890588     -0.005653   -0.172714            -0.208516  
31428              -0.332086      2.053177    0.964079            -0.208516  
...                      ...           ...         ...                  ...  
26896              -0.332086     -0.432693   -0.294513            -0.208516  
17365              -0.332086      0.142408    0.040435            -0.208516  
53692              -0.332086      0.020842   -0.091514            -0.208516  
57525               2.113262      0.924794    0.426132            -0.208516  
36996              -0.332086     -0.657122   -0.426462            -0.208516  

[14455 rows x 11 columns]